File: misc/002-Add Two Numbers/main-recursive.c
#define NULL 0

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* addTwoNumbersRecurse(struct ListNode* l1, struct ListNode* l2, int carry);

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2)
{
    // Base case
    // If L1 is empty, then L2 is either also empty, or contains solution.
    if(l1 == NULL)
        return l2;
    // If L2 is empty then L1 contains the solution.
    if(l2 == NULL)
        return l1;

    struct ListNode *n1, *n2;

    n1 = malloc(sizeof(struct ListNode));
    n2 = malloc(sizeof(struct ListNode));
    n1->next = l1;
    n2->next = l2;

    return addTwoNumbersRecurse(l1, l2, 0);

}

struct ListNode* addTwoNumbersRecurse(struct ListNode* l1, struct ListNode* l2, int carry)
{
    int v1, v2, vTotal, vRemainder;
    struct ListNode* rNode;

    if((l1->next == NULL) && (l2->next == NULL))
    {
        if(carry)
        {
            rNode = malloc(sizeof(struct ListNode));
            rNode->val = carry;
            rNode->next = NULL;
        }
        else
        {
            return NULL;
        }
    }

    // Initialize working data
    v1 = v2 = 0;
    if(l1->next != NULL)
    {
        v1 = l1->next->val;
        l1 = l1->next;
    }

    if(l2->next != NULL)
    {
        v2 = l2->next->val;
        l2 = l2->next;
    }

    // Calculate node values
    vTotal = v1 + v2 + carry;
    vRemainder = vTotal % 10;
    carry = vTotal / 10;

    // Instanticiate new node
    rNode = malloc(sizeof(struct ListNode));
    rNode->val = vRemainder;
    rNode->next = addTwoNumbersRecurse(l1, l2, carry);
    return rNode;
}

File: misc/002-Add Two Numbers/main-recursive.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1, l2):

        # Default exit case
        # If L1 is empty, then L2 is either empty, or is the return value.
        if(l1 == None):
            return l2
        # If L2 is empty then L1 was not empty, and is the return value.
        if(l2 == None):
            return l1

        t1 = ListNode(val=0,next=l1)
        t2 = ListNode(val=0,next=l2)
        return self.addTwoNumbersRecurse(t1,t2,0)

    def addTwoNumbersRecurse(self, l1, l2, carry):

        # Base case
        if(l1.next is None and l2.next is None):
            return ListNode(val=carry) if carry is not 0 else None

        vL = 0 if l1.next is None else l1.next.val
        vR = 0 if l2.next is None else l2.next.val
        vTotal = vL + vR + carry
        vRemainder = vTotal % 10
        carry = vTotal // 10
        
        t1 = l1.next if l1.next is not None else l1
        t2 = l2.next if l2.next is not None else l2
        return ListNode(vRemainder,self.addTwoNumbersRecurse(t1, t2, carry))


if(__name__ == "__main__"):

    n1 = [9,9,9]
    n2 = [8,9,9,9,9,9]

    l1 = ListNode()
    w = l1
    for v in n1:
        w.next = ListNode(val=v)
        w = w.next

    l2 = ListNode()
    w = l2
    for v in n2:
        w.next = ListNode(val=v)
        w = w.next
    
    l1 = l1.next
    l2 = l2.next   

    w = Solution().addTwoNumbers(l1,l2)
    while(w is not None):
        print(w.val)
        w = w.next